TheWarWithin/RogueAssassination.lua

```diff
@@ -2096,24 +2102,38 @@ spec:RegisterAbilities( {
         cooldown = 0,
         gcd = "totem",
         school = "physical",
-
         spend = 35,
         spendType = "energy",
 
         startsCombat = true,
         cycle = function () return buff.deadly_poison.up and "deadly_poison_dot" or buff.amplifying_poison.up and "amplifying_poison_dot" or nil end,
 
+        cp_gain = function()
+            local fanCP = buff.clear_the_witnesses.up and 2 or 1
+            
+            -- Predict crit gains
+            if talent.seal_fate.enabled then
+                -- calculate the crit chance of Fan of Knives then estimate 
+                fanCP = fanCP + floor( true_active_enemies * ( 0.01 * ( crit_pct_current + ( talent.deadly_precision.enabled and 5 or 0 ) + ( talent.thrown_precision.enabled and 5 or 0 ) + ( buff.momentum_of_despair.up and 10 or 0 ) + ( buff.master_assassin_any.up and 20 or 0 ) ) ) )
+            end
+
+            return fanCP
+        end,
+
         handler = function ()
-            gain( buff.clear_the_witnesses.up and 2 or 1, "combo_points" )
+            gain( action.fan_of_knives.cp_gain, "combo_points" )
             removeBuff( "hidden_blades" )
             removeBuff( "clear_the_witnesses" )
 
+            local newPoisons = floor( poison_chance * true_active_enemies )
+
             if buff.deadly_poison.up then
                 applyDebuff( "target", "deadly_poison_dot" )
-                active_dot.deadly_poison_dot = min( active_enemies, active_dot.deadly_poison_dot + 8 )
-            elseif buff.amplifying_poison.up then
+                active_dot.deadly_poison_dot = min( active_enemies, active_dot.deadly_poison_dot + newPoisons )
+            end
+            if buff.amplifying_poison.up then
                 applyDebuff( "target", "amplifying_poison_dot" )
-                active_dot.amplifying_poison_dot = min( active_enemies, active_dot.amplifying_poison_dot + 8 )
+                active_dot.amplifying_poison_dot = min( active_enemies, active_dot.deadly_poison_dot + newPoisons )
```

Comment:
I think you mean `active_dot.amplifiying_poison_dot` here inside the `min` call.

Reply:
This calculation is also questionable. We can assume that the target gets a the lethal poison dot even though it's only a 30ish% chance, but just adding the number of possible new poison dots to the number of existing lethal poison dots seems like a high likelihood of overcounting. Should we do more bookkeeping to make this estimate more accurate?

Reply:
The correct calculation is:
```
local newDeadlyPoisons = floor( poison_chance * max( 0, true_active_enemies - active_dot.deadly_poison_dot ) )
active_dot.deadly_poison_dot = min( active_enemies, active_dot.deadly_poison_dot + newDeadlyPoisons )
```
And then similarly for Amplifying Poison.

Reply:
> The correct calculation is:
> 
> ```
> local newDeadlyPoisons = floor( poison_chance * max( 0, true_active_enemies - active_dot.deadly_poison_dot ) )
> active_dot.deadly_poison_dot = min( active_enemies, active_dot.deadly_poison_dot + newDeadlyPoisons )
> ```
> 
> And then similarly for Amplifying Poison.

I'm not sure the difference is significant given that you don't know whether the existing dots are on the active enemies, or the enemies slightly out of range.

> I think you mean active_dot.amplifiying_poison_dot here inside the min call.

Yep!

Reply:
In that case, I'd at least document that the calculation is only an approximation and why in case we have to revisit this code again.

Reply:
Agreed, I will do that